### PR TITLE
fix: normalize FUNCTION LOAD library code for shebang/newlines (#3307)

### LINKED ISSUE #3307: ResponseError: Missing library metadata
**Version**: redis-5.0.7
**Platform**: Python 3.9  on Windows 10
**Description**: 

docs: https://redis.io/docs/latest/commands/function-load/

## redis-cli is fine
create redis from docker
```
docker run -d --name redis_ts3 -p 16379:6379 redis/redis-stack-server
docker exec -it redis_ts3 /bin/bash && redis-cli
```

Add function without error
```
127.0.0.1:6379> FUNCTION LOAD "#!lua name=mylib \n redis.register_function('myfunc', function(keys, args) return args[1] end)"
"mylib"
127.0.0.1:6379>
```

## redis-py raise error

```python
import redis

settings = {
    'host': 'localhost',
    'port': 16379,
    'db': 0,
    'password': None,
}

r = redis.Redis(**settings)
r.function_load("#!lua name=mylib \n redis.register_function('myfunc', function(keys, args) return args[1] end)")
```

error:

```
ResponseError                             Traceback (most recent call last)
Cell In[9], line 11
      3 settings = {
      4     'host': 'localhost',
      5     'port': 16379,
      6     'db': 0,
      7     'password': None,
      8 }
     10 r = redis.Redis(**settings)
---> 11 r.function_load("#!lua name=mylib \n redis.register_function('myfunc', function(keys, args) return args[1] end)")

File c:\envs\quant\lib\site-packages\redis\commands\core.py:6014, in FunctionCommands.function_load(self, code, replace)
   6012 pieces = ["REPLACE"] if replace else []
   6013 pieces.append(code)
-> 6014 return self.execute_command("FUNCTION LOAD", *pieces)

File c:\envs\quant\lib\site-packages\redis\client.py:545, in Redis.execute_command(self, *args, **options)
    542 conn = self.connection or pool.get_connection(command_name, **options)
    544 try:
--> 545     return conn.retry.call_with_retry(
    546         lambda: self._send_command_parse_response(
    547             conn, command_name, *args, **options
    548         ),
    549         lambda error: self._disconnect_raise(conn, error),
    550     )
    551 finally:
    552     if not self.connection:

File c:\envs\quant\lib\site-packages\redis\retry.py:46, in Retry.call_with_retry(self, do, fail)
     44 while True:
     45     try:
---> 46         return do()
     47     except self._supported_errors as error:
     48         failures += 1

File c:\envs\quant\lib\site-packages\redis\client.py:546, in Redis.execute_command.<locals>.<lambda>()
    542 conn = self.connection or pool.get_connection(command_name, **options)
    544 try:
    545     return conn.retry.call_with_retry(
--> 546         lambda: self._send_command_parse_response(
    547             conn, command_name, *args, **options
    548         ),
    549         lambda error: self._disconnect_raise(conn, error),
    550     )
    551 finally:
    552     if not self.connection:

File c:\envs\quant\lib\site-packages\redis\client.py:522, in Redis._send_command_parse_response(self, conn, command_name, *args, **options)
    518 """
    519 Send a command and parse the response
    520 """
    521 conn.send_command(*args)
--> 522 return self.parse_response(conn, command_name, **options)

File c:\envs\quant\lib\site-packages\redis\client.py:562, in Redis.parse_response(self, connection, command_name, **options)
    560         options.pop(NEVER_DECODE)
    561     else:
--> 562         response = connection.read_response()
    563 except ResponseError:
    564     if EMPTY_RESPONSE in options:

File c:\envs\quant\lib\site-packages\redis\connection.py:536, in AbstractConnection.read_response(self, disable_decoding, disconnect_on_error, push_request)
    534 if isinstance(response, ResponseError):
    535     try:
--> 536         raise response
    537     finally:
    538         del response  # avoid creating ref cycles

ResponseError: Library 'mylib' already exists
```

### PR DESCRIPTION
**Pull request by sankalpsthakur**
Fixes #3307.

## Summary

Normalize `FUNCTION LOAD` source before sending it:

- remove leading whitespace before the shebang, including Unicode whitespace in `str` inputs
- expand only an escaped shebang terminator
- normalize CRLF, LFCR, and lone CR shebang endings to LF
- preserve body bytes, body escapes, and trailing whitespace
- keep unchanged contiguous binary inputs zero copy
- scan text and binary framing linearly, avoiding regex backtracking on large terminator-free inputs

## Tests

- 22 focused helper tests pass locally
- Redis integration coverage includes indented and redis-cli style sources
- Ruff lint and format checks pass

AI tools assisted with the change. I reviewed and tested the final implementation.

---

> [!NOTE]
> **Low Risk**
> Scoped to FUNCTION LOAD payload shaping with extensive tests; no auth, cluster routing, or general command-encoding changes.
> 
> **Overview**
> **`function_load`** now runs library source through **`normalize_function_lib_code`** so pasted redis-cli/docs examples load without Redis metadata errors.
> 
> The helper strips leading whitespace (including Unicode on `str`), expands only the **first** shebang-line terminator when it is a literal `\n`/`\r\n` escape or CR/CRLF, and normalizes that terminator to LF while leaving the Lua body (line endings, escapes, trailing space) unchanged. **`str`**, **`bytes`**, **`bytearray`**, and **`memoryview`** are supported; unchanged contiguous binary inputs can be returned without copying.
> 
> Docs on **`function_load`** describe the behavior. Unit tests cover edge cases (backslash runs, non-contiguous views, round trips); integration tests cover indented triple-quoted strings and redis-cli-style escaped shebangs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1e8a1b1c0627aa1d4c0a908664741ccac099006c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by sankalpsthakur** [comment]
Addressed the Codex review on `normalize_function_lib_code`:

When a single-line payload has no real newline, we now expand **only the first** redis-cli-style shebang terminator (`\r\n` or `\n` escape), not every `\n` in the payload. Later sequences (e.g. Lua `return 'a\nb'`) are left intact.

Unit coverage added for single-line body `\n` preservation and first-only `\r\n` escape expansion. Helper tests: 12 passed.

File: redis/commands/helpers.py [hunk 1]
```diff
@@ -201,3 +201,48 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+def _normalize_function_lib_code_text(text: str) -> str:
+    # Drop leading/trailing whitespace so the shebang is at offset 0.
+    text = text.strip()
+    # Normalize common line-ending variants to LF (what Redis's parser scans for).
+    text = text.replace("\r\n", "\n").replace("\r", "\n")
+    # Expand redis-cli-style "\n" escapes only when no real newline is present,
+    # so multi-line Lua that intentionally contains a backslash-n sequence is
+    # left alone.
+    if "\n" not in text and "\\n" in text:
+        text = text.replace("\\n", "\n")
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in 1171d86c: single-line payloads now expand only the first shebang terminator escape (`\r\n` / `\n`), so intentional Lua body escapes like `return 'a\nb'` are preserved.

**Comment by sankalpsthakur** [comment]
Remaining Codex/Bugbot points addressed in e17d595a:
- Expand first shebang terminator escape even when later real newlines exist
- Choose first terminator by position of `\r\n` vs `\n`
- Bytes path uses latin-1 so non-UTF-8 bodies still get framing fixes

**Comment by sankalpsthakur** [comment]
Addressed the remaining review findings in `4e56d1d`: binary framing now stays bytes-native with a no-copy unchanged path, trailing whitespace is preserved, and regressions cover both. All review threads are resolved. The new CI runs are awaiting maintainer approval.

**Comment by sankalpsthakur** [comment]
All current review findings are addressed through `b7d7c24`: normalization changes only the shebang terminator and preserves Lua body line endings byte-for-byte, including a physical body CR after an escaped CRLF; escaped LF still absorbs a following CR as the LFCR terminator. Contiguous memoryview slices and foreign views retain identity when unchanged, while unsupported non-contiguous views copy only when required. Focused text/bytes assertions pass. All current review threads are resolved. GitHub workflows are awaiting maintainer approval.

File: redis/commands/helpers.py [hunk 2]
```diff
@@ -201,3 +202,113 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+    esc_crlf = text.find("\\r\\n", 0, header_end)
+    esc_lf = text.find("\\n", 0, header_end)
+    escaped = min((pos for pos in (esc_crlf, esc_lf) if pos >= 0), default=-1)
+    if escaped >= 0:
+        length = 4 if escaped == esc_crlf else 2
+        text = text[:escaped] + "\n" + text[escaped + length :]
+
+    # Lua treats CRLF and LFCR as one newline. Expand first so an escaped
+    # terminator followed by CR is normalized as a single LFCR newline.
+    return re.sub(r"\r\n|\n\r|\r", "\n", text)
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in `0daf8b5`: normalization now rewrites only the shebang terminator. Lua body CRLF, LFCR, and lone-CR bytes are preserved exactly, with a focused text/bytes regression.

File: redis/commands/helpers.py [hunk 3]
```diff
@@ -201,3 +202,113 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+            view = searchable = code
+        else:
+            # A full view over bytes/bytearray can use the backing object's
+            # C-level find without copying. Copy sliced/foreign views once.
+            backing = view.obj
+            if isinstance(backing, (bytes, bytearray)) and view.nbytes == len(backing):
+                searchable = backing
+            else:
+                code = bytes(view)
+                view = code
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in `0daf8b5`: contiguous memoryviews, including slices and foreign backing objects, are inspected in place and returned by identity when framing is already valid. Only unsupported non-contiguous view shapes copy, and the focused regression asserts the no-copy slice path.

File: redis/commands/helpers.py [hunk 4]
```diff
@@ -201,3 +201,139 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+    real_nl = min((pos for pos in (real_lf, real_cr) if pos >= 0), default=-1)
+    header_end = real_nl if real_nl >= 0 else len(text)
+    esc_crlf = text.find("\\r\\n", 0, header_end)
+    esc_lf = text.find("\\n", 0, header_end)
+    escaped = min((pos for pos in (esc_crlf, esc_lf) if pos >= 0), default=-1)
+
+    if escaped >= 0:
+        terminator_end = escaped + (4 if escaped == esc_crlf else 2)
+        if terminator_end < len(text) and text[terminator_end] == "\r":
+            terminator_end += 1
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed through `b7d7c24`: only an escaped `\n` absorbs a following physical `\r` as an LFCR shebang terminator. An escaped `\r\n` is already complete, so a following physical CR is now preserved as Lua body content. Added focused text and bytes regressions for both cases.

File: tests/test_helpers.py [hunk 5]
```diff
@@ -47,3 +48,187 @@ def test_random_string():
...
+    assert normalize_function_lib_code(text.encode()) == expected.encode()
+
+
+@pytest.mark.fixed_client
+def test_normalize_function_lib_code_preserves_body_cr_after_escaped_crlf():
+    text = r"#!lua name=mylib \\r\\n" + "\rreturn 1"
+    expected = "#!lua name=mylib \n\rreturn 1"
+
+    assert normalize_function_lib_code(text) == expected
+    assert normalize_function_lib_code(text.encode()) == expected.encode()
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply tests/test_helpers.py:194]: Fixed in 4e5a054. The fixture now uses the intended raw `\r\n` sequence. The focused helper suite passes.

File: redis/commands/helpers.py [hunk 6]
```diff
@@ -201,3 +201,139 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+    real_cr = _find_function_lib_bytes(view, b"\r", start)
+    real_nl = min((pos for pos in (real_lf, real_cr) if pos >= 0), default=-1)
+    header_end = real_nl if real_nl >= 0 else len(view)
+    esc_crlf = _find_function_lib_bytes(view, b"\\r\\n", start, header_end)
+    esc_lf = _find_function_lib_bytes(view, b"\\n", start, header_end)
+    escaped = min((pos for pos in (esc_crlf, esc_lf) if pos >= 0), default=-1)
+
+    valid_lf = real_lf >= 0 and (real_cr < 0 or real_lf < real_cr)
+    if start == 0 and escaped < 0 and (valid_lf or real_nl < 0):
+        return code
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in 4e5a054. Text and binary paths now both collapse CRLF and LFCR shebang terminators to one LF, including after leading whitespace. Added a text and bytes regression.

File: redis/commands/core.py [hunk 7]
```diff
@@ -12283,16 +12283,24 @@ def function_load(self, code: str, replace: bool | None = False) -> (
...
-        :param code: the source code (must start with
-        Shebang statement that provides a metadata about the library)
+
+        :param code: the source code. It must include a shebang that provides
+            metadata about the library (``#!lua name=<lib>``). Leading
+            whitespace is stripped so multi-line strings work; trailing
+            whitespace is preserved for exact source round trips. If the code
+            is a single line that uses a redis-cli-style ``\\n`` escape
+            after the shebang instead of a real newline, that first terminator
+            escape is expanded so Redis can parse library metadata. Later
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/core.py]: Fixed in 4e5a054. The docstring now describes an escaped shebang terminator and explicitly covers payloads whose body contains later physical newlines.

File: redis/commands/helpers.py [hunk 8]
```diff
@@ -201,3 +201,171 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+            and terminator_end < len(text)
+            and text[terminator_end] == "\r"
+        ):
+            terminator_end += 1
+        return text[:escaped] + "\n" + text[terminator_end:]
+
+    valid_lf = (
+        real_lf >= 0
+        and (real_cr < 0 or real_lf < real_cr)
+        and (real_lf + 1 == len(text) or text[real_lf + 1] != "\r")
```
Comment (Sanjays2402) [inline redis/commands/helpers.py]: the LFCR clause misfires when the shebang already ends in a plain \n and the body starts with a CRLF blank line: "#!lua name=mylib\n\r\nreturn 1" fails valid_lf (next char is \r), so the \r gets swallowed as part of an LFCR terminator and the body comes back "\n\nreturn 1". ran it against the patch to confirm. body-preserved-byte-for-byte does not hold there; treating a lone \n as a valid terminator regardless of what follows would fix it.

Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in `d647f79`: a plain LF now always ends the shebang, so a following CRLF stays in the body. Added text and bytes regressions for this exact case.

**Comment by sankalpsthakur** [comment]
Addressed the two remaining performance findings in `c04497f`. Binary shebang framing now uses one early-stop scan for physical or escaped terminators. Changed `bytes` and `memoryview` payloads are assembled with a single final allocation, while `bytearray` uses one exact-size buffer. A focused local matrix passed, including a 5 MB escaped-terminator memoryview, unchanged-view identity, non-contiguous fallback, and CR/LF body preservation. The two review threads are resolved; upstream workflows are waiting for fork authorization.

File: redis/commands/core.py [hunk 9]
```diff
@@ -12283,16 +12283,23 @@ def function_load(self, code: str, replace: bool | None = False) -> (
...
-        Shebang statement that provides a metadata about the library)
+
+        :param code: the source code. It must include a shebang that provides
+            metadata about the library (``#!lua name=<lib>``). Leading
+            whitespace is stripped so multi-line strings work; trailing
+            whitespace is preserved for exact source round trips. If the
+            shebang terminator is a redis-cli-style ``\\n`` or ``\\r\\n``
+            escape instead of a real newline, that terminator is expanded even
+            when the body contains later physical newlines. Later escapes in
+            the body (e.g. Lua string escapes) are left unchanged.
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/core.py]: Agreed. The intended public wording is: physical CRLF and lone-CR shebang terminators are normalized to LF, while later body escapes and physical line endings remain unchanged. The PR summary and helper contract now disclose this behavior; the source-method wording still needs the same sentence.

**Comment by sankalpsthakur** [comment]
Remaining documentation-only item: update `FunctionCommands.function_load` to state that physical CRLF and lone-CR shebang terminators are normalized to LF, while later Lua body escapes and physical line endings remain unchanged. Implementation and focused behavior coverage are complete.

**Comment by sankalpsthakur** [comment]
Addressed the two fresh helper findings in `610664f9`: text framing now uses a linear scanner instead of a backtracking regex, and leading `str` whitespace follows `str.isspace()` semantics. Added focused regressions for a 50,000-character terminator-free input and non-ASCII leading whitespace. Both helper threads are resolved. The remaining open thread is documentation-only in `FunctionCommands.function_load`; the intended wording is already captured in the PR summary.

**Comment by sankalpsthakur** [comment]
The remaining documentation item is now in the code as well as the summary: `e763941c` extends the `FunctionCommands.function_load` docstring to state that physical CRLF and lone-CR shebang terminators are normalized to LF while the body's own physical line endings are preserved. No behavioral change in this commit; docstring only.

File: redis/commands/helpers.py [hunk 10]
```diff
@@ -201,3 +201,174 @@ def at_most_one_value_set(iterable: Iterable[Any]):
...
+                and data[pos + 1] == ord("r")
+                and data[pos + 2] == ord("\\")
+                and data[pos + 3] == ord("n")
+            ):
+                return pos, pos + 4, True
+            if pos + 1 < size and data[pos + 1] == ord("n"):
+                end = pos + 2
+                if end < size and data[end] == ord("\r"):
+                    end += 1
+                return pos, end, True
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/helpers.py]: Fixed in a0298670: both the binary and text terminator scanners now measure the backslash run length and only treat an odd run's final backslash as the escape start, so a literal `\n` following paired backslashes is treated as content. Added regressions for the even-run identity case (text and bytes), the odd-run case, and `\\r\n` parity. Full local helper suite: 27/27 normalize tests passing.

File: redis/commands/core.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply redis/commands/core.py]: Documented in e763941c: the `function_load` docstring now states that physical CRLF and lone-CR shebang terminators are normalized to LF, while the body's own physical line endings are preserved.

**Comment by petyaslavova** [comment]
Hey @sankalpsthakur, thank you for your contribution! I'm starting the review :)

File: tests/test_helpers.py [hunk 11]
```diff
@@ -47,3 +48,228 @@ def test_random_string():
...
+    assert isinstance(normalized, bytes)
+    assert normalized == code
+
+
+@pytest.mark.fixed_client
+def test_normalize_function_lib_code_odd_backslash_run_still_terminates():
+    # Three backslashes: the first two are literal, the third starts `\n`.
+    code = "#!lua name=mylib " + "\\\\\\n" + " body"
+    normalized = normalize_function_lib_code(code)
+    assert normalized == "#!lua name=mylib " + "\\\\" + "\n body"
```
[the remark this answers is not in the document]
Comment (sankalpsthakur) [reply tests/test_helpers.py]: Fixed in `1e8a1b1c`: the fixture now constructs three literal backslashes plus `n` explicitly, asserts the input differs from the expected normalized output, and exercises the odd-run terminator path without ambiguous Python escaping.

**Comment by petyaslavova** [comment]
Hey @sankalpsthakur, thanks for the detailed work in this PR, and for iterating through the review rounds.

After checking the details in #3307, I see that the failing payload is a triple-quoted Python string that begins with a newline, and Redis documents that the FUNCTION LOAD payload must start with the` #! `shebang.
Sending the source verbatim is the correct client behavior — `r.function_load(code.strip())` loads it. 
The reporter's other form, "#!lua name=mylib \n ...", already contains a real newline in Python and works today; tests/test_function.py::test_function_load has covered exactly that spelling for years. 
The redis-cli \n is unescaped by redis-cli's own argument parser, so a Python caller only produces the two-character form deliberately, and no report shows that case.

Normalizing the payload client-side would make redis-py accept input the command contract declares invalid, behave differently from execute_command("FUNCTION LOAD", ...), and stop FUNCTION LIST WITHCODE returning the source that was passed. I'd rather not take on that permanent contract, so I'm going to close this PR. 

The better approach to address #3307 would be to clarify in the function_load docstring that the source must begin with #! with no leading whitespace — if you'd like to send that change instead, it would be very welcome.
